trie/smt.py

```diff
@@ -0,0 +1,355 @@
+        return tuple(self._branch)
+
+    @property
+    def root_hash(self) -> Hash32:
+        return calc_root(self.key, self.value, self.branch)
+
+    def merge(self, key: bytes, value: bytes, node_updates: Sequence[Hash32]):
+        """
+        Merge an update for another key with the one we are tracking internally.
+
+        :param key: keypath of the update we are processing
+        :param value: value of the update we are processing
+        :param node_updates: sequence of siblings (in root->leaf order)
+                             must be at least as large as the first diverging
+                             key in the keypath
+
+        """
+        validate_is_bytes(key)
+        validate_length(key, self._key_size)
+
+        # Path diff is the logical XOR of the updated key and this account
+        path_diff = (to_int(self.key) ^ to_int(key))
+
+        # Same key (diff of 0), update the tracked value
+        if path_diff == 0:
+            self._value = value
+            # No need to update branch
+        else:
+            # Find the first mismatched bit between keypaths. This is
+            # where the branch point occurs, and we should update the
+            # sibling node in the source branch at the branch point.
+            # NOTE: Keys are in MSB->LSB (root->leaf) order.
+            #       Node lists are in root->leaf order.
+            #       Be sure to convert between them effectively.
+            for bit in reversed(range(self._branch_size)):
+                if path_diff & (1 << bit) > 0:
+                    branch_point = (self._branch_size-1)-bit
+                    break
+
+            # NOTE: node_updates only has to be as long as necessary
+            #       to obtain the update. This allows an optimization
+            #       of pruning updates to the maximum possible depth
+            #       that would be required to merge, which may be many
+            #       orders of magnitude smaller than the tree depth.
+            if len(node_updates) <= branch_point:
+                raise ValidationError("Updated node list is not deep enough")
+
+            # Update sibling in the branch where our key differs from the update
+            self._branch[branch_point] = node_updates[branch_point]
+            # No need to update value
+
+
+class SparseMerkleTree:
+
+    def __init__(self, key_size: int=32, default: bytes=BLANK_NODE):
+        # Ensure we can support the given depth
+        if not 1 <= key_size <= 32:
+            raise ValidationError("Keysize must be number of bytes in range [0, 32]")
+
+        self._key_size = key_size  # key size in bytes
+        self.depth = key_size * 8  # depth is number of bits in key
+
+        self._default = default
+
+        # Initialize an empty tree with one branch
+        self.db = {}
+        node = self._default  # Default leaf node
+        for _ in range(self.depth):
+            node_hash = keccak(node)
+            self.db[node_hash] = node
```

Comment:
One optimization (probably to go in an issue for later): it looks like these node hashes could be stored in a blank hash ancestor lookup so the `get()` can shortcut and return a blank result as soon as it hits one of the blank hash ancestors. (making it `O(log(h))` instead of `O(h)` operations to find a blank key)

Reply:
I think we can leave this out for now, but definitely something to think about.